src/algebra/module_regular.lean

```diff
@@ -0,0 +1,168 @@
+/-
+Copyright (c) 2021 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+import algebra.group
+import algebra.group_power.basic
+import algebra.iterate_hom
+import algebra.module.basic
+
+/-!
+# Action of regular elements on a module
+
+We introduce `M`-regular elements, in the context of an `R`-module `M`.
+
+By definition, an `M`-regular element of a commutative ring acting on an `R`-module `M` is an
+element `a ∈ R` such that the function `m ↦ a • m` is injective.
+-/
+
+--should it be in the module namespace?
+
+namespace module
+
+variables {R : Type*} {a b : R} (M : Type*)
+
+section has_scalar
+
+variables [has_scalar R M]
+
+/-- An `M`-regular element is an element `c` such that multiplication on the left by `c` is an
+injective map `M → M`. -/
+def is_regular (c : R) := function.injective ((•) c : M → M)
+
+section monoid
+
+variables [monoid R] [add_monoid M] [is_scalar_tower R R M]
```

Comment:
Isn't `is_scalar_tower R R M` another way to spell `mul_action R M`?

Reply:
I think what I'm actually advocating is that this say `is_scalar_tower R S M` then, and you deal with the special case of `mul_action R M` below. Perhaps we're missing `mul_action.is_scalar_tower.left` or similar?

Reply:
As I was proving these lemmas, I also had the impression that there were some missing instances and `mul_action` and `is_scalar_tower` do not seem to interact as well as they could (or at least as well as they do in my mind).

Reply:
In any case, you are right about the scalar_tower thing: I will change the "middle" `R` to a separate ring and prove that statement instead.

Reply:
The instance I'm thinking of exists and is `is_scalar_tower.left`. Perhaps you just don't have it imported, or perhaps it belongs in a different place.